### PR TITLE
Refactor handle error handler

### LINKED ISSUE #16294: Error.handler.ts and api.interceptor.ts might be customizeble/overridable
For an example , user can handle spesific errors.
defaults must be moved to seperate file. 
every function must be overridable so on...

### PR DESCRIPTION
**Pull request by mahmut-gundogdu**
### Description

Resolves https://github.com/abpframework/abp/issues/16294

### Checklist

- [x] I fully tested it as developer / designer and created unit / integration tests
- [x] I documented it (or no need to document or I will create a separate documentation issue)

### How to test it?
When you run the app, everything should work.

### CODE REVIEW

File: npm/ng-packs/packages/theme-shared/src/lib/constants/error.ts
Comment (masum-ulu) [inline npm/ng-packs/packages/theme-shared/src/lib/constants/error.ts:1]: Maybe file name can be the `default-errors.ts` It's up to you

File: npm/ng-packs/packages/theme-shared/src/lib/constants/error.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  },
+  defaultError500: {
+    title: 'AbpUi::500Message',
+    details: 'AbpUi::DefaultErrorMessage',
+  },
+};
+
+export const CUSTOM_HTTP_ERROR_HANDLER_PRIORITY = Object.freeze({
+  veryLow: -9,
+  low: -9,
```
Comment (masum-ulu) [inline npm/ng-packs/packages/theme-shared/src/lib/constants/error.ts]: Why both have same value ?

File: npm/ng-packs/packages/theme-shared/src/lib/services/create-error-component.service.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import { ResolveEnd } from '@angular/router';
+import { filter } from 'rxjs/operators';
+import { RouterEvents } from '@abp/ng.core';
+import { CanCreateCustomErrorService } from './can-create-custom-error.service';
+import { HTTP_ERROR_CONFIG } from '../tokens/http-error.token';
+import { HttpErrorWrapperComponent } from '../components/http-error-wrapper/http-error-wrapper.component';
+import { ErrorScreenErrorCodes } from '../models/common';
+
+@Injectable({ providedIn: 'root' })
+export class CreateErrorComponentService {
```
Comment (masum-ulu) [inline npm/ng-packs/packages/theme-shared/src/lib/services/create-error-component.service.ts:20]: This and other services needs to have signs

File: npm/ng-packs/packages/theme-shared/src/lib/services/can-create-custom-error.service.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { inject, Injectable } from '@angular/core';
+import { ErrorScreenErrorCodes } from '../models/common';
+import { HTTP_ERROR_CONFIG } from '../tokens/http-error.token';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class CanCreateCustomErrorService {
```
Comment (masum-ulu) [inline npm/ng-packs/packages/theme-shared/src/lib/services/can-create-custom-error.service.ts]: This service's method can move to **CreateErrorComponentService**

File: npm/ng-packs/packages/theme-shared/src/lib/services/tenant-resolve-error-handler.service.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import { CustomHttpErrorHandlerService } from '../models/common';
+import { inject, Injectable } from '@angular/core';
+import { AuthService } from '@abp/ng.core';
+import { HttpErrorResponse } from '@angular/common/http';
+import { CUSTOM_HTTP_ERROR_HANDLER_PRIORITY } from '../constants/error';
+
+@Injectable({ providedIn: 'root' })
+export class TenantResolveErrorHandlerService implements CustomHttpErrorHandlerService {
+  priority = CUSTOM_HTTP_ERROR_HANDLER_PRIORITY.high;
```
Comment (mahmut-gundogdu) [inline npm/ng-packs/packages/theme-shared/src/lib/services/tenant-resolve-error-handler.service.ts]: it must be readonly

File: npm/ng-packs/packages/theme-shared/src/lib/services/unknown-status-code-error-handler.service.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  CUSTOM_HTTP_ERROR_HANDLER_PRIORITY,
+  DEFAULT_ERROR_LOCALIZATIONS,
+  DEFAULT_ERROR_MESSAGES,
+} from '../constants/error';
+import { inject, Injectable } from '@angular/core';
+import { CreateErrorComponentService } from './create-error-component.service';
+
+@Injectable({ providedIn: 'root' })
+export class UnknownStatusCodeErrorHandlerService implements CustomHttpErrorHandlerService {
+  priority = CUSTOM_HTTP_ERROR_HANDLER_PRIORITY.normal;
```
Comment (mahmut-gundogdu) [inline npm/ng-packs/packages/theme-shared/src/lib/services/unknown-status-code-error-handler.service.ts]: readonly

File: npm/ng-packs/packages/theme-shared/src/lib/theme-shared.module.ts [hunk 6]
```diff
@@ -144,6 +143,7 @@ export class ThemeSharedModule {
             ...(confirmationIcons || {}),
           },
         },
+        errorHandlersProviders,
```
Comment (mahmut-gundogdu) [inline npm/ng-packs/packages/theme-shared/src/lib/theme-shared.module.ts]: rename the variable. It must be uppercase

File: npm/ng-packs/packages/theme-shared/src/lib/handlers/error.handler.ts [hunk 7]
```diff
@@ -1,128 +1,45 @@
...
+    this.routerErrorHandlerService.listen();
   }
 
   protected listenToRestError() {
     this.httpErrorReporter.reporter$
       .pipe(filter(this.filterRestErrors), switchMap(this.executeErrorHandler))
-      .subscribe();
+      .subscribe(err => {
+        this.handleError(err);
+      });
```
Comment (masum-ulu) [inline npm/ng-packs/packages/theme-shared/src/lib/handlers/error.handler.ts:39]: Just suggestion

```ts
.subscribe(this.handleError)
```

File: npm/ng-packs/packages/theme-shared/src/lib/models/common.ts [hunk 8]
```diff
@@ -21,9 +21,12 @@ export interface HttpErrorConfig {
   };
 }
 
-export type HttpErrorHandler = (
-  injector: Injector,
-  httpError: HttpErrorResponse,
-) => Observable<any>;
+export type HttpErrorHandler = (httpError: HttpErrorResponse) => Observable<any>;
```
Comment (masum-ulu) [inline npm/ng-packs/packages/theme-shared/src/lib/models/common.ts]: What do you think about 🤔
```ts
export type HttpErrorHandler<T = any> = (httpError: HttpErrorResponse) => Observable<T>;
```

File: npm/ng-packs/packages/theme-shared/src/lib/services/create-error-component.service.ts (hunk 2 above)
Comment (mahmut-gundogdu) [reply npm/ng-packs/packages/theme-shared/src/lib/services/create-error-component.service.ts:20]: Like what? Interface ?

File: npm/ng-packs/packages/theme-shared/src/lib/models/common.ts (hunk 8 above)
Comment (mahmut-gundogdu) [reply npm/ng-packs/packages/theme-shared/src/lib/models/common.ts]: Better

**Review by hikalkan — APPROVED** [review]

**Review by masum-ulu — APPROVED** [review]

**Pull request merged**